README.md
# entangle

entangle 取至量子纠缠(Quantum Entanglement)

entangle 是基于 websocket 运行在 dart VM 的服务端框架

在 flutter 中, 与之配套的有 entangle-client 可以在客户端更简洁的调用 entangle, 并且处理好了断线重连， 心跳重连等机制

entangle-client 配合 entangle 能让我们像在本机异步处理对象一样访问服务器的函数, 并传递 json 对象

其他非 dart 语言的客户端亦可以直接使用 websocket 进行调用

## Server:

```dart
import 'package:entangle/entangle.dart';

class Context {
  String name = 'i am a context data';
}

void main() async {
  var entangle = Entangle({
    'hello-world': hello,
  }, context: Context());

  entangle.listen(port: 5000, path: '/ws');
}

// controller example
void hello(Context ctx, data, send) {
  send({'msg': '${data['msg']}-world', 'ctx-name': ctx.name});

  // if other == 100, send message at 1000ms ago
  if (data['other'] == 100) {
    Future.delayed(Duration(milliseconds: 1000), () {
      send({'other': 'server take the initiative at 1000ms ago'});
    });
  }
}

```

## client:

```dart
import 'package:entangle/entangle-client.dart';

void clientApp() async {
  // wait server runing
  await Future.delayed(Duration(milliseconds: 500), () {});

  await entangleClient.connect('ws://127.0.0.1:5000/ws');

  // heart check at 1500ms, if connect.close, auto conect again;
  entangleClient.connectHeart(1500);

  /// Simple send data, and wait server callback twice
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloA",
      "other": 100,
    },
    listen: (data, remove) {
      print('client-get-msg-1: $data');
    },
  );

  ///  Only run once listen ad every send();
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloB",
      "other": false,
    },
    listen: (data, remove) {
      print('client-get-msg-2: $data');

      // remove this listen
      remove();
    },
  );

  /// timeOut Feature
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloB",
      "other": false,
    },
    listen: (data, clear) {
      print('client-get-msg-2: $data');
    },
    timeOutMilliseconds: 5000,
  );

  /// Auto connect again
  ///
  /// Example:
  Future.delayed(Duration(milliseconds: 1500), () async {
    print('close');
    await entangleClient.client.close();

    // Auto connect at entangleClient.send;
    entangleClient.send(
      controller: 'hello-world',
      data: {"msg": "helloA"},
      listen: (data, clear) {
        print('client-get-msg-again: $data');
      },
    );
  });
}
```


CHANGELOG.md
## [0.1.1] - Change API

Change API:

old: send(String controller, Map data, {Function listen})
new: send({String controller, Map data, Function listen})

## [0.1.0] - Update readme

## [0.0.2] - Make Features

## [0.0.1] - init

- TODO: Describe initial release.


lib/entangle.dart
library entangle;

import 'dart:convert';
import 'dart:io';

String _entangleType = '__entangle__';
// String _entangleKey = 'k';
String _entangleController = 'c';

class Entangle {
  final Map<String, Function> controllers;
  dynamic context;

  Entangle(this.controllers, {this.context}) {
    if (this.controllers == null) {
      throw 'Entangle need init params: controllers';
    }
  }

  Future listen({
    int port = 5000,
    String path = '/ws',
    Function(HttpRequest) handle,
  }) async {
    var server = await HttpServer.bind(InternetAddress.loopbackIPv4, port);
    print('entangle running: http:127.0.0.1:$port');
    print('entangle-ws running: ws:127.0.0.1:$port$path');
    await for (HttpRequest req in server) {
      // 如果路径相同，就使用 websocket
      if (req.uri.path == path) {
        serveRequest(req).catchError((Error error) {
          print('WebSocket error: --start');
          print(error);
          print(error.stackTrace);
          print('[WebSocket error] --end');
        });
      } else {
        if (handle != null) {
          handle(req);
        } else {
          req.response.close();
        }
      }
    }
  }

  Future serveRequest(HttpRequest request) {
    if (WebSocketTransformer.isUpgradeRequest(request)) {
      return WebSocketTransformer.upgrade(request).then((webSocket) {
        // webSockets.add(webSocket);
        webSocket.listen((dynamic msg) {
          handleMessage(msg, webSocket);
        });
      });
    } else {
      request.response.statusCode = HttpStatus.notAcceptable;
      request.response.writeln('Need use WebSocket connect');
      request.response.close();
      return new Future(() {});
    }
  }

  Function webSocketListen;

  void handleMessage(dynamic msg, WebSocket webSocket) {
    if (webSocket.closeCode == null) {
      Map<String, dynamic> data;
      try {
        data = json.decode(msg);
      } catch (err) {
        print('[Server Error], webSocket msg can not decode: $msg');
        if (webSocketListen != null) {
          webSocketListen(msg);
        }
      }

      // if data have _entangleType;
      if (data != null && data.containsKey(_entangleType)) {
        Map<String, dynamic> theType = data[_entangleType];

        if (theType.containsKey(_entangleController)) {
          String ctrl = theType[_entangleController].split('.')[0];
          if (controllers.containsKey(ctrl)) {
            Function ctrlFn = controllers[ctrl];

            void send(Map<String, dynamic> reback) {
              if (webSocket.closeCode == null) {
                Map<String, dynamic> rebackMap = {};
                rebackMap.addAll(reback);
                rebackMap[_entangleType] = theType;
                webSocket.add(json.encode(rebackMap));
              }
            }

            ctrlFn(context, data, send);
          }
        }
      }
    }
  }
}

List<String> entangleGetPaths(Map<String, dynamic> data) {
  if (data.containsKey(_entangleType)) {
    Map<String, dynamic> theType = data[_entangleType];

    if (theType.containsKey(_entangleController)) {
      List<String> paths = theType[_entangleController].split('.');
      return paths;
    }
    return null;
  }
  return null;
}


lib/entangle-client.dart
import 'dart:io';
import 'dart:convert';

var entangleClient = _EntangleClient();

String _entangleType = '__entangle__';
String _entangleKey = 'k';
String _entangleController = 'c';

class _EntangleClient {
  int sendReConnectMilliseconds = 500;
  int autoConnectCount = 5;
  int _autoConnectingCount = 0;
  int _key = 0;
  Map<String, Function> _sendCallbacks = {};

  String lastConnectUrl = '';

  WebSocket client;

  void connectHeart(int heartMilliseconds) {
    Future.delayed(Duration(milliseconds: heartMilliseconds), () {
      if (client.readyState != 1) {
        connect(lastConnectUrl);
      } else {
        connectHeart(heartMilliseconds);
      }
    });
  }

  Future connect(String url) async {
    if (lastConnectUrl == '') {
      lastConnectUrl = url;
    }
    client = await WebSocket.connect(url);
    _autoConnectingCount = 0;
    client.listen(_onListen);
    return client;
  }

  void _add(/*String|List<int>*/ data) {
    Future.delayed(Duration(milliseconds: 4), () {
      if (client.readyState == 1) {
        client.add(data);
      } else if (client.readyState != 1) {
        _autoConnectingCount += 1;
        if (_autoConnectingCount < autoConnectCount) {
          Future.delayed(Duration(milliseconds: sendReConnectMilliseconds),
              () async {
            await connect(lastConnectUrl);
            _add(data);
          });
        } else {
          print('$lastConnectUrl connect is error!');
        }
      }
    });
  }

  int _updateSendKey() {
    _key += 1;
    if (_key > 999999) {
      _key = 0;
    }
    return _key;
  }

  Function send({
    String controller,
    Map<String, dynamic> data,
    Function listen,
    int timeOutMilliseconds = 8000,
  }) {
    int theKey = _updateSendKey();
    if (!controller.contains('.')) {
      controller = controller + '.*';
    }

    if (listen != null) {
      _sendCallbacks[theKey.toString()] = listen;
    }

    data[_entangleType] = {};
    data[_entangleType][_entangleController] = controller;
    data[_entangleType][_entangleKey] = theKey;
    String obj = json.encode(data);
    _add(obj);

    void removeListen() {
      _sendCallbacks.remove(theKey.toString());
    }

    if (timeOutMilliseconds > 0) {
      Future.delayed(Duration(milliseconds: timeOutMilliseconds), removeListen);
    }

    return removeListen;
  }

  void _onListen(dynamic msg) async {
    Map<String, dynamic> data;
    try {
      data = json.decode(msg);
    } catch (err) {
      print('[Client Error], WebSocket msg can not decode: $msg');
    }

    if (data.containsKey(_entangleType)) {
      void removeListen() {
        _sendCallbacks.remove(data[_entangleType][_entangleKey].toString());
      }

      _sendCallbacks[data[_entangleType][_entangleKey].toString()](
          data, removeListen);
    }
  }

  Future close([int code, String reason]) async {
    await client.close(code, reason);
  }
}


example/example-client.dart
import 'package:entangle/entangle-client.dart';

void clientApp() async {
  // wait server runing
  await Future.delayed(Duration(milliseconds: 500), () {});

  await entangleClient.connect('ws://127.0.0.1:5000/ws');

  // heart check at 1500ms, if connect.close, auto conect again;
  entangleClient.connectHeart(1500);

  /// Simple send data, and wait server callback twice
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloA",
      "other": 100,
    },
    listen: (data, remove) {
      print('client-get-msg-1: $data');
    },
  );

  ///  Only run once listen ad every send();
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloB",
      "other": false,
    },
    listen: (data, remove) {
      print('client-get-msg-2: $data');

      // remove this listen
      remove();
    },
  );

  /// timeOut Feature
  ///
  /// Example:
  entangleClient.send(
    controller: 'hello-world',
    data: {
      "msg": "helloB",
      "other": false,
    },
    listen: (data, clear) {
      print('client-get-msg-2: $data');
    },
    timeOutMilliseconds: 5000,
  );

  /// Auto connect again
  ///
  /// Example:
  Future.delayed(Duration(milliseconds: 1500), () async {
    print('close');
    await entangleClient.client.close();

    // Auto connect at entangleClient.send;
    entangleClient.send(
      controller: 'hello-world',
      data: {"msg": "helloA"},
      listen: (data, clear) {
        print('client-get-msg-again: $data');
      },
    );
  });
}
